### PR TITLE
[language/tools] Add Move code coverage summaries, and source coverage

### PR DESCRIPTION
**Pull request by tzakian**
This PR adds a couple things:
* Coverage reporting of a module (and/or the stdlib) based off of a trace that was collected from the VM (usually during a full test).
* Source reporting of coverage (see screenshot below). _Note that we don't support multiline non-coverage yet_.

## Coverage Reports
An example call passing in a serialized coverage map `coverage_map.mvcov` (created via the `move-trace-conversion` binary), and checking the coverage of the standard library:

```
cargo run -p move-coverage --bin coverage-summaries -- -t coverage_map.mvcov -s ../../stdlib/staged/stdlib.mv
```

This will result in the following output:

```
+-------------------------+
| Move Coverage Summary   |
+-------------------------+
Module 00000000000000000000000000000000::Debug
>>> % Module coverage: NaN
Module 00000000000000000000000000000000::Testnet
>>> % Module coverage: 80.000
Module 00000000000000000000000000000000::LibraTimestamp
>>> % Module coverage: 74.648
Module 00000000000000000000000000000000::Association
>>> % Module coverage: 95.161
Module 00000000000000000000000000000000::AccountLimits
>>> % Module coverage: 36.190
Module 00000000000000000000000000000000::Vector
>>> % Module coverage: 100.000
Module 00000000000000000000000000000000::Empty
>>> % Module coverage: 100.000
Module 00000000000000000000000000000000::AccountType
>>> % Module coverage: 95.238
Module 00000000000000000000000000000000::VASP
>>> % Module coverage: 85.595
Module 00000000000000000000000000000000::Unhosted
>>> % Module coverage: 58.824
Module 00000000000000000000000000000000::AccountTrack
>>> % Module coverage: 71.765
Module 00000000000000000000000000000000::Signature
>>> % Module coverage: NaN
Module 00000000000000000000000000000000::LibraTransactionTimeout
>>> % Module coverage: 64.815
Module 00000000000000000000000000000000::Offer
>>> % Module coverage: 100.000
Module 00000000000000000000000000000000::LCS
>>> % Module coverage: NaN
Module 00000000000000000000000000000000::Event
>>> % Module coverage: 88.028
Module 00000000000000000000000000000000::LibraConfig
>>> % Module coverage: 90.094
Module 00000000000000000000000000000000::RegisteredCurrencies
>>> % Module coverage: 90.625
Module 00000000000000000000000000000000::FixedPoint32
>>> % Module coverage: 100.000
Module 00000000000000000000000000000000::Libra
>>> % Module coverage: 90.985
Module 00000000000000000000000000000000::Hash
>>> % Module coverage: NaN
Module 00000000000000000000000000000000::LibraAccount
>>> % Module coverage: 92.899
Module 00000000000000000000000000000000::ApprovedPayment
>>> % Module coverage: 94.118
Module 00000000000000000000000000000000::Authenticator
>>> % Module coverage: 96.739
Module 00000000000000000000000000000000::Coin1
>>> % Module coverage: 100.000
Module 00000000000000000000000000000000::Coin2
>>> % Module coverage: 100.000
Module 00000000000000000000000000000000::Compare
>>> % Module coverage: 100.000
Module 00000000000000000000000000000000::Option
>>> % Module coverage: 97.015
Module 00000000000000000000000000000000::ValidatorConfig
>>> % Module coverage: 87.342
Module 00000000000000000000000000000000::LibraSystem
>>> % Module coverage: 91.193
Module 00000000000000000000000000000000::TransactionFee
>>> % Module coverage: 87.368
Module 00000000000000000000000000000000::LibraWriteSetManager
>>> % Module coverage: 93.103
Module 00000000000000000000000000000000::LBR
>>> % Module coverage: 88.426
Module 00000000000000000000000000000000::LibraBlock
>>> % Module coverage: 94.444
Module 00000000000000000000000000000000::Genesis
>>> % Module coverage: 100.000
Module 00000000000000000000000000000000::LibraVMConfig
>>> % Module coverage: 100.000
Module 00000000000000000000000000000000::LibraVersion
>>> % Module coverage: 81.250
Module 00000000000000000000000000000000::PaymentRouter
>>> % Module coverage: 93.678
Module 00000000000000000000000000000000::SharedEd25519PublicKey
>>> % Module coverage: 100.000
+-------------------------+
| % Move Coverage: 87.528 |
+-------------------------+
```

You can also turn on coverage reporting for each function as well by passing the `-f` flag:
```
+-------------------------+
| Move Coverage Summary   |
+-------------------------+
Module 00000000000000000000000000000000::Debug
        native fun print
                total instructions: 0
                covered instructions: 0
                % coverage: NaN
        native fun print_stack_trace
                total instructions: 0
                covered instructions: 0
                % coverage: NaN
>>> % Module coverage: NaN
Module 00000000000000000000000000000000::Testnet
        fun initialize
                total instructions: 12
                covered instructions: 9
                % coverage: 75.000
        fun is_testnet
                total instructions: 3
                covered instructions: 3
                % coverage: 100.000
>>> % Module coverage: 80.000
Module 00000000000000000000000000000000::LibraTimestamp
        fun initialize
                total instructions: 12
                covered instructions: 9
                % coverage: 75.000
        fun is_genesis
                total instructions: 4
                covered instructions: 4
                % coverage: 100.000
        fun now_microseconds
                total instructions: 5
                covered instructions: 5
                % coverage: 100.000
        fun update_global_time
                total instructions: 50
                covered instructions: 35
                % coverage: 70.000
>>> % Module coverage: 74.648
```

## Source Code Coverage

You can also use the coverage map along with the source map, compiled module, and its source file to get a source-level coverage report of the Move code:
```
 cargo run -p move-coverage --bin source-coverage -- -t coverage_map.mvcov -b <path to compiled module> -s <path to source file>
``` 
An example from running this on the association module:
Screen Shot 2020-05-09 at 3 14 27 PM

### CODE REVIEW

File: language/tools/move-coverage/src/summary.rs [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+                    fn_summary.total_number_of_instructions
+                )?;
+                writeln!(
+                    summary_writer,
+                    "\t\tcovered instructions: {}",
+                    fn_summary.covered_instructions
+                )?;
+                writeln!(
+                    summary_writer,
+                    "\t\t% coverage: {:.3}",
```
Comment (sblackshear) [inline language/tools/move-coverage/src/summary.rs:131]: Nit: One or two decimal places seems like enough here/might be more readable

File: language/tools/move-coverage/src/bin/source-coverage.rs [hunk 2]
```diff
@@ -0,0 +1,74 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+use bytecode_source_map::utils::{remap_owned_loc_to_loc, source_map_from_file, OwnedLoc};
+use move_coverage::{coverage_map::CoverageMap, source_coverage::SourceCoverageBuilder};
+//use move_ir_types::location::Spanned;
```
Comment (sblackshear) [inline language/tools/move-coverage/src/bin/source-coverage.rs:8]: Delete?

File: language/tools/move-coverage/src/bin/coverage-summaries.rs [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+use structopt::StructOpt;
+use vm::file_format::CompiledModule;
+
+#[derive(Debug, StructOpt)]
+#[structopt(
+    name = "Move VM Coverage Summary",
+    about = "Creates a coverage summary from the trace data collected from the Move VM"
+)]
+struct Args {
+    /// The path to the coverage map or trace filee
```
Comment (dariorussi) [inline language/tools/move-coverage/src/bin/coverage-summaries.rs:24]: file or files

Comment (tzakian) [reply language/tools/move-coverage/src/bin/coverage-summaries.rs:24]: It should be 'file' Updating now :)

**Pull request closed without merging**